File: shared-libraries/go-logging/logger.go
package logging

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/metric"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
	"go.opentelemetry.io/otel/trace"
)

// Logger handles structured logging, metrics, and tracing
type Logger struct {
	serviceName     string
	version         string
	environment     string
	tracer          trace.Tracer
	meter           metric.Meter
	requestCounter  metric.Int64Counter
	requestDuration metric.Float64Histogram
	initialized     bool
}

// Config holds the configuration for the logger
type Config struct {
	ServiceName string
	Version     string
	Environment string
	AlloyURL    string
}

// New creates a new logger instance
func New(config Config) *Logger {
	logger := &Logger{
		serviceName: config.ServiceName,
		version:     config.Version,
		environment: config.Environment,
	}

	// Initialize OpenTelemetry if AlloyURL is provided
	if config.AlloyURL != "" {
		logger.initOpenTelemetry(config.AlloyURL)
	}

	return logger
}

// initOpenTelemetry sets up OpenTelemetry components
func (l *Logger) initOpenTelemetry(alloyURL string) {
	ctx := context.Background()

	// Create resource with service information
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(l.serviceName),
			semconv.ServiceVersion(l.version),
			semconv.DeploymentEnvironment(l.environment),
		),
	)
	if err != nil {
		log.Printf("Failed to create resource: %v", err)
		return
	}

	// Initialize tracing
	l.initTracing(ctx, res, alloyURL)

	// Initialize metrics
	l.initMetrics(ctx, res, alloyURL)

	l.initialized = true
}

// initTracing sets up tracing
func (l *Logger) initTracing(ctx context.Context, res *resource.Resource, alloyURL string) {
	// Create OTLP trace exporter
	traceExporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(alloyURL),
		otlptracehttp.WithInsecure(),
	)
	if err != nil {
		log.Printf("Failed to create trace exporter: %v", err)
		return
	}

	// Create trace provider
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
	)

	// Set global trace provider
	otel.SetTracerProvider(tp)

	// Create tracer
	l.tracer = tp.Tracer(l.serviceName)
}

// initMetrics sets up metrics
func (l *Logger) initMetrics(ctx context.Context, res *resource.Resource, alloyURL string) {
	// Create OTLP metric exporter
	metricExporter, err := otlpmetrichttp.New(ctx,
		otlpmetrichttp.WithEndpoint(alloyURL),
		otlpmetrichttp.WithInsecure(),
	)
	if err != nil {
		log.Printf("Failed to create metric exporter: %v", err)
		return
	}

	// Create meter provider
	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExporter)),
	)

	// Set global meter provider
	otel.SetMeterProvider(mp)

	// Create meter
	l.meter = mp.Meter(l.serviceName)

	// Create metrics
	l.requestCounter, err = l.meter.Int64Counter(
		"http_requests_total",
		metric.WithDescription("HTTP requests"),
	)
	if err != nil {
		log.Printf("Failed to create http_requests_total counter: %v", err)
	}

	l.requestDuration, err = l.meter.Float64Histogram(
		"http_request_duration_seconds",
		metric.WithDescription("Request duration in seconds"),
	)
	if err != nil {
		log.Printf("Failed to create http_request_duration_seconds histogram: %v", err)
	}
}

// Logging functions

// Info logs an info message
func (l *Logger) Info(ctx context.Context, message string, fields ...map[string]interface{}) {
	l.log(ctx, "INFO", message, fields...)
}

// Error logs an error message
func (l *Logger) Error(ctx context.Context, message string, err error, fields ...map[string]interface{}) {
	allFields := []map[string]interface{}{{"error": err.Error()}}
	allFields = append(allFields, fields...)
	l.log(ctx, "ERROR", message, allFields...)
}

// Warn logs a warning message
func (l *Logger) Warn(ctx context.Context, message string, fields ...map[string]interface{}) {
	l.log(ctx, "WARN", message, fields...)
}

// Debug logs a debug message
func (l *Logger) Debug(ctx context.Context, message string, fields ...map[string]interface{}) {
	l.log(ctx, "DEBUG", message, fields...)
}

// log is the internal logging function
func (l *Logger) log(ctx context.Context, level, message string, fields ...map[string]interface{}) {
	logData := map[string]interface{}{
		"timestamp":   time.Now().UTC().Format(time.RFC3339),
		"level":       level,
		"message":     message,
		"service":     l.serviceName,
		"version":     l.version,
		"environment": l.environment,
	}

	// Add trace context automatically
	if l.initialized && l.tracer != nil {
		if span := trace.SpanFromContext(ctx); span.IsRecording() {
			logData["trace_id"] = span.SpanContext().TraceID().String()
			logData["span_id"] = span.SpanContext().SpanID().String()
		}
	}

	// Merge all fields
	for _, fieldMap := range fields {
		for k, v := range fieldMap {
			logData[k] = v
		}
	}

	// Send to stdout (will be collected by Loki)
	jsonData, _ := json.Marshal(logData)
	log.Println(string(jsonData))
}

// Metric functions

// CountRequest increments the request counter
func (l *Logger) CountRequest(ctx context.Context, endpoint string, statusCode int) {
	if l.initialized && l.requestCounter != nil {
		l.requestCounter.Add(ctx, 1, metric.WithAttributes(
			attribute.String("endpoint", endpoint),
			attribute.String("status_code", fmt.Sprintf("%d", statusCode)),
			attribute.String("service", l.serviceName),
		))
	}
}

// RecordDuration records request duration
func (l *Logger) RecordDuration(ctx context.Context, endpoint string, duration time.Duration) {
	if l.initialized && l.requestDuration != nil {
		l.requestDuration.Record(ctx, duration.Seconds(), metric.WithAttributes(
			attribute.String("endpoint", endpoint),
			attribute.String("service", l.serviceName),
		))
	}
}

// Tracing functions

// StartSpan starts a new span
func (l *Logger) StartSpan(ctx context.Context, operation string) (context.Context, func()) {
	if l.initialized && l.tracer != nil {
		ctx, span := l.tracer.Start(ctx, operation)
		span.SetAttributes(
			attribute.String("service", l.serviceName),
			attribute.String("version", l.version),
			attribute.String("environment", l.environment),
		)

		return ctx, func() {
			span.End()
		}
	}

	// Return no-op if not initialized
	return ctx, func() {}
}

// AddSpanEvent adds an event to the current span
func (l *Logger) AddSpanEvent(ctx context.Context, event string, fields ...map[string]interface{}) {
	if l.initialized && l.tracer != nil {
		if span := trace.SpanFromContext(ctx); span.IsRecording() {
			attrs := []attribute.KeyValue{
				attribute.String("event", event),
			}

			// Convert fields to attributes
			for _, fieldMap := range fields {
				for k, v := range fieldMap {
					attrs = append(attrs, attribute.String(k, fmt.Sprintf("%v", v)))
				}
			}

			span.AddEvent(event, trace.WithAttributes(attrs...))
		}
	}
}

// AddSpanAttribute adds an attribute to the current span
func (l *Logger) AddSpanAttribute(ctx context.Context, key, value string) {
	if l.initialized && l.tracer != nil {
		if span := trace.SpanFromContext(ctx); span.IsRecording() {
			span.SetAttributes(attribute.String(key, value))
		}
	}
}

File: applications/user-service/shared-libraries/python-logging/faidon_laboratory_logging/logger.py
import json
import time
from typing import Dict, Any, Optional
from dataclasses import dataclass

from opentelemetry import trace, metrics


class DummySpan:
    """Dummy span that acts as a context manager when OpenTelemetry is not initialized"""
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        pass
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.semconv.resource import ResourceAttributes


@dataclass
class Config:
    """Configuration for the logger"""
    service_name: str
    version: str
    environment: str
    alloy_url: Optional[str] = None


class Logger:
    """Structured logging with OpenTelemetry integration"""
    
    def __init__(self, config: Config):
        self.service_name = config.service_name
        self.version = config.version
        self.environment = config.environment
        self.initialized = False
        
        # Initialize OpenTelemetry if AlloyURL is provided
        if config.alloy_url:
            self._init_opentelemetry(config.alloy_url)
    
    def _init_opentelemetry(self, alloy_url: str):
        """Initialize OpenTelemetry components"""
        try:
            # Create resource
            resource = Resource.create({
                ResourceAttributes.SERVICE_NAME: self.service_name,
                ResourceAttributes.SERVICE_VERSION: self.version,
                ResourceAttributes.DEPLOYMENT_ENVIRONMENT: self.environment,
            })
            
            # Initialize tracing
            self._init_tracing(resource, alloy_url)
            
            # Initialize metrics
            self._init_metrics(resource, alloy_url)
            
            self.initialized = True
            
        except Exception as e:
            print(f"Failed to initialize OpenTelemetry: {e}")
    
    def _init_tracing(self, resource: Resource, alloy_url: str):
        """Initialize tracing"""
        try:
            # Create OTLP trace exporter
            trace_exporter = OTLPSpanExporter(
                endpoint=f"http://{alloy_url}/v1/traces",
            )
            
            # Create tracer provider
            tracer_provider = TracerProvider(resource=resource)
            tracer_provider.add_span_processor(
                BatchSpanProcessor(trace_exporter)
            )
            
            # Set global tracer provider
            trace.set_tracer_provider(tracer_provider)
            
            # Create tracer
            self.tracer = trace.get_tracer(self.service_name)
            
        except Exception as e:
            print(f"Failed to initialize tracing: {e}")
            self.tracer = None
    
    def _init_metrics(self, resource: Resource, alloy_url: str):
        """Initialize metrics"""
        try:
            # Create OTLP metric exporter
            metric_exporter = OTLPMetricExporter(
                endpoint=f"http://{alloy_url}/v1/metrics",
            )
            
            # Create meter provider
            meter_provider = MeterProvider(
                resource=resource,
                metric_readers=[
                    PeriodicExportingMetricReader(metric_exporter)
                ]
            )
            
            # Set global meter provider
            metrics.set_meter_provider(meter_provider)
            
            # Create meter
            self.meter = metrics.get_meter(self.service_name)
            
            # Create metrics
            self.request_counter = self.meter.create_counter(
                name="http_requests_total",
                description="HTTP requests"
            )
            
            self.request_duration = self.meter.create_histogram(
                name="http_request_duration_seconds",
                description="Request duration in seconds"
            )
            
        except Exception as e:
            print(f"Failed to initialize metrics: {e}")
            self.request_counter = None
            self.request_duration = None
    
    # Logging functions
    
    def info(self, message: str, **fields):
        """Log an info message"""
        self._log("INFO", message, fields)
    
    def error(self, message: str, error: Exception, **fields):
        """Log an error message"""
        fields["error"] = str(error)
        self._log("ERROR", message, fields)
    
    def warn(self, message: str, **fields):
        """Log a warning message"""
        self._log("WARN", message, fields)
    
    def debug(self, message: str, **fields):
        """Log a debug message"""
        self._log("DEBUG", message, fields)
    
    def _log(self, level: str, message: str, fields: Dict[str, Any]):
        """Internal logging function"""
        log_data = {
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "level": level,
            "message": message,
            "service": self.service_name,
            "version": self.version,
            "environment": self.environment,
            **fields
        }
        
        # Add trace context automatically
        if self.initialized:
            span = trace.get_current_span()
            if span and span.is_recording():
                span_context = span.get_span_context()
                log_data["trace_id"] = format(span_context.trace_id, '032x')
                log_data["span_id"] = format(span_context.span_id, '016x')
        
        # Send to stdout (will be collected by Loki)
        print(json.dumps(log_data))
    
    # Metric functions
    
    def count_request(self, endpoint: str, status_code: int):
        """Increment the request counter"""
        if self.initialized and self.request_counter:
            self.request_counter.add(1, {
                "endpoint": endpoint,
                "status_code": str(status_code),
                "service": self.service_name
            })
    
    def record_duration(self, endpoint: str, duration_seconds: float):
        """Record request duration"""
        if self.initialized and self.request_duration:
            self.request_duration.record(duration_seconds, {
                "endpoint": endpoint,
                "service": self.service_name
            })
    
    # Tracing functions
    
    def start_span(self, operation: str):
        """Start a new span"""
        if self.initialized and self.tracer:
            span = self.tracer.start_span(operation)
            span.set_attributes({
                "service": self.service_name,
                "version": self.version,
                "environment": self.environment
            })
            return span
        # Return a dummy context manager if not initialized
        return DummySpan()
    
    def add_span_event(self, event: str, **fields):
        """Add an event to the current span"""
        if self.initialized:
            span = trace.get_current_span()
            if span and span.is_recording():
                span.add_event(event, fields)
    
    def add_span_attribute(self, key: str, value: str):
        """Add an attribute to the current span"""
        if self.initialized:
            span = trace.get_current_span()
            if span and span.is_recording():
                span.set_attribute(key, value)
